### ISSUE TITLE
hard limits on time range ( controlled by operator?)

### ISSUE THREAD

**Issue by Dieterbe** [grafana/grafana#4288]
currently there's a quick range of 5years back, but using the calendar picker and zoom out controls, you can go arbitrarily far back (e.g. 30 or 100 years even)

however:
* the tsdb may only have data up to a certain retention (a year, 3 months, ... ?) so it doesn't make sense to go back further
* performance could degrade significantly or even tsdb crash if you go too far back.

so i propose admin controls to:
1) limit how far back one can go
2) what the max time range is

this would result in disabling of certain quick range options, restricting how far "zoom out" can go, and shielding users from certain timeranges in the calendar picker/editor.

we could perhaps tie this in to the quota config system because available retentions can be dependent on which org you're in.

**Comment by lucadistefano** [comment]
+1

**Comment by gsueur** [comment]
+1

**Comment by CarusDave** [comment]
+1

A feature like this would be very useful. We have some "time focused" dashboards using lots of singlestat panels where we ideally need to restrict/disable the time range selection.

Having the ability to override the user time selection using a admin dashboard setting would be great. The benefits are two fold; keeps panels showing expected current data and stops unnecessary data retrival.

**Comment by theotheredbaron** [comment]
This would be really useful for us as well - we have some time-critical monitoring where there is a lot of data but only the last 30 mins is really relevant.  We've already had issues with users clicking "last year" and it hammering the DB.

**Comment by IvanVas** [comment]
+1
Use case: publicly available dashboard

**Comment by kb1lqc** [comment]
+1 also using a publicly available dashboard

**Comment by Exordian** [comment]
+1 this is a blocker for publicly available dashboards

**Comment by biohazardxxx** [comment]
+1

**Comment by oleg-glushak** [comment]
+1

**Comment by Brausmith** [comment]
Yes please.
I would specifically like to 'lock' single stats so that they always show a time period. This would allow the user to change the dashboard time period but still have these single stats always show data for exact time periods, like last 7 days, 30 days, or 1 year.
While not as important, I would also like to do this with graphs.

**Comment by rtd** [comment]
+1

**Comment by Vish17agr** [comment]
+1
When we can expect this feature. I can see feature request has been opened in 2016.
Is there any side way to limit the user to not provide large time range in Dashboard.

**Comment by AndresPineros** [comment]
This would be the cherry on top of the cake for Grafana. I'm very worried that someone who may not understand the implications of big time ranges could render Grafana useless for other users.

**Comment by gpurrenhage** [comment]
+1

**Comment by klau2005** [comment]
+1

**Comment by sna696** [comment]
+1

**Comment by bobbydoogle** [comment]
+1

**Comment by reinhard-brandstaedter** [comment]
+1

**Comment by jackpgao** [comment]
+1

**Comment by pacoweb** [comment]
+1

**Comment by atanasio** [comment]
+1

**Comment by bsmarcosj** [comment]
+1

**Comment by petersonbasso** [comment]
+1

**Comment by bassmaster187** [comment]
+1

**Comment by mahaveer1707** [comment]
True. This feaure is a must.

+1

**Comment by davddcb** [comment]
+1

**Comment by AdrianBarcelo** [comment]
+1

**Comment by ashishkaransingh** [comment]
It is imperative to have an ability to Limit or lock "Time-range/Date-picker" specially for High Resolution data.
+1

Thanks

**Comment by ITGuyDave** [comment]
I run Grafana on a RPI for IoT sensor monitoring and graphing (low power server essentially). If one runs a search any further back than a week, the load average shoots through the roof and goes unresponsive for many minutes or hours. I cannot see why there is no restriction on this setting currently, given a much larger system than this could render it unusable in multi-user environments.

This isn't so much a performance concern to Grafana - it's more a worry to the underlying database impact by a design decision made by Grafana and users not understanding the implications.

Please consider this feature!

**Comment by torkelo** [comment]
It's because Grafana was created for fast time series databases (where most initially had rollups) and long time range queries where super quick. But it makes a lot more sense now with the diverse set of data sources Grafana supports.

**Comment by bsmarcosj** [comment]
@torkelo I don't think that the problem is long time range queries, but the fact that to do a query with a big range of time that causes to the BD retrieve a lot of data... i.e. "last 7 days", "last 30 days", etc.

**Comment by torkelo** [comment]
@bsmarcosj that is usually not a problem if your query is constructed correctly. Long time range should 
 not mean progressively more data returned to Grafana. That means your query is missing a group by time interval or some other way to specify max data points/date interval grouping & aggregation.

**Comment by Dieterbe** [comment]
This argument seems to assume 1) good intent by all grafana users 2) thoughtful, careful crafting of each query.
I argue this premise doesn't apply for the majority of grafana deployments.
Even with graphite, if you hit the coarsest rollups, with grafana you can go so many years back that the number of points still becomes substantial. (correct me if i'm wrong, but IIRC if you ask graphite for data 100 years back and you only have 2 years of data, you'll get 98 years worth of nulls in additional to your runtime-consolidated data from the coarsest archive)

**Comment by sadeychai** [comment]
+1

**Comment by mashuma** [comment]
Absolute +1

For our use case, this is a huge inconvenience.
We have specialised dashboards, one example:
- Data retention is 2 weeks
- Intended view range is from now-1day to now (1 day)

Ability to go beyond the intended time or even data retention period is confusing and counter-intuitive.
Currently, we keep enabled "Hide time picker", to try to avoid user selecting invalid range with no data. But this is by far not perfect, as this also removes the user to be able to set appropriate refresh, sub-range within allowed range etc.

This is much needed feature.

**Comment by brian15co** [comment]
+1 

This is something that would greatly enhance the application's utility

**Comment by bbcelly** [comment]
+1

**Comment by baur** [comment]
It would be useful for slow datasourses ...

**Comment by zagaria** [comment]
+1

**Comment by CMahesh** [comment]
+1 from myside. I was search for this features. Since, for the enterprise products the query over the huge time range will be a issue.

**Comment by Dieterbe** [comment]
@CMahesh with enterprise products are you referring to?

**Comment by ayomide101** [comment]
+1

**Comment by CMahesh** [comment]
@Dieterbe In Enterprise Manager of our company we have huge data. We don't want the people to query after certain range.

**Comment by Nemesarial** [comment]
+1 Please -- extended time ranges puts our thanos-store under crushing pressure

**Comment by oded-dd** [comment]
Is there a plan to implement it in the foreseeable future?

**Comment by sebas-w** [comment]
+1, Zooming out to a period like 2 years crashes the TSDB as it consumes all memory to respond to grafana dashboard when we have high resolution data.

**Comment by ITGuyDave** [comment]
> 
> 
> This argument seems to assume 1) good intent by all grafana users 2) thoughtful, careful crafting of each query.
> I argue this premise doesn't apply for the majority of grafana deployments.
> Even with graphite, if you hit the coarsest rollups, with grafana you can go so many years back that the number of points still becomes substantial. (correct me if i'm wrong, but IIRC if you ask graphite for data 100 years back and you only have 2 years of data, you'll get 98 years worth of nulls in additional to your runtime-consolidated data from the coarsest archive)
> 
> I guess one could argue each TSDB itself should protect itself from bad requests that would cause an overload. That's the direction we took with metrictank, at least.

@Dieterbe - I was just reading about metrictank. Based on this, are you able to elaborate if this is also the case with platforms such as InfluxDB (in reference to the nulls)? In InfluxDB I can't go back more than 30 days or it chews up all RAM + swap, even at a coarse level. I noticed you are the lead dev on metrictank, and being able to run such large queries without essentially chewing up memory in the way InfluxDB does (even while under low load) could actually make these types of queries more viable on smaller platforms too. I would happily switch to metrictank even if it's immature if it meant a more usable platform on smaller devices...

**Comment by Dieterbe** [comment]
I can't comment on InfluxDB. Haven't used it in a long time.
Metrictank is designed for clusters running on servers (possibly VM's or containers). It's not meant for small devices.

**Comment by oholimoli** [comment]
It is really necessary to have this feature. It could be a preliminary solution to adjust the $__timefilter variables in order to prevent to long selections by setting a $maxtimerange variable in the dashboard settings.

**Comment by jwhite878** [comment]
+1, we can't have users running arbitrarily large queries at scale.

**Comment by ptsilva** [comment]
+1

**Comment by maxless** [comment]
Almost four years and still nothing? C'mon guys, making a hard limit on period length in the dashboard options shouldn't be that hard.

**Comment by theotheredbaron** [comment]
We gave up waiting and used views to limit the time period.  Grafana is free software and I'm very grateful for it - if I had the coding skills I'd try to fix it myself!

**Comment by maxless** [comment]
> We gave up waiting and used views to limit the time period. Grafana is free software and I'm very grateful for it - if I had the coding skills I'd try to fix it myself!

Do you mean panel time ranges or something else?

**Comment by dmned-ph** [comment]
I use a work around by creating a constant variable named TimeLimit. The value can be 3d, 7d, 30d, etc., for example.

Then I modify the queries by replacing **$timeFilter** with **time >= now() - $TimeLimit**.

**Comment by serge-salamanka-1pt** [comment]
+1, trying to create a dashboard with performance test results for specific time range that cannot be changed.

**Comment by sbaygithub** [comment]
Please add this feature. We need it for larger dashboards.

**Comment by dattasoumya92** [comment]
+1 Please add this feature, this can be useful in limiting date time range to a max range (say last 3 months).

**Comment by pencilcap123** [comment]
+1 Please add this feature, this can be useful in limiting date time range to a max range (say last 3 months).

We beg you !

**Comment by ahlaw** [comment]
I'm thinking of taking a crack at implementing this as additional dashboard settings. Should the additional settings also enforce [Panel time overrides](https://grafana.com/docs/grafana/latest/dashboards/time-range-controls/#panel-time-overrides-and-timeshift) and [URL controls](https://grafana.com/docs/grafana/latest/dashboards/time-range-controls/#control-the-time-range-using-a-url)?

**Comment by tollk** [comment]
+1.
From my point of view URL controls yes
But Panel time overrides not as this is only available to editors.

**Comment by CommanderPirx** [comment]
+1 Please add this (simple) feature. Without it I can't make a public-facing data display without a risk of revealing too much data.

**Comment by misteio** [comment]
+1 ...
5 years later, and no newz ?

**Comment by aurelien-clu** [comment]
+1

**Comment by sniff122** [comment]
+1 i really would like to have a public dashboard for my discord bot stats but this is my main issue of doing that due to the large amount of records you could fetch

**Comment by shhsiao** [comment]
+1...

**Comment by edgardbraz** [comment]
Can someone review this PR? Or the latest versions have support to this kind of configuration to preserve resources used unnecessarily?

**Comment by MukherjeeJeet** [comment]
#36384

**Comment by kamakazikamikaze** [comment]
#36384 appears to have been closed due to a permissions issue in regards to branches, so #47055 has been opened in its stead. It appears that @MukherjeeJeet is not actively working on this PR. Is there anyone else who could see it through?

**Comment by tguenneguez** [comment]
👍

**Comment by alvarobacelar** [comment]
+1 This feature will be very userfull for larger dashboards.

**Comment by ibo2** [comment]
+1 must have to avoid crashes due to users big selections

**Comment by Wyle09** [comment]
+1 we are facing the same issue that @ibo2 stated

**Comment by dgsardina** [comment]
+1 to avoid users selecting long time intervals and hammering the db

**Comment by diegoscarabelli** [comment]
What's the plan?

**Comment by harisbal** [comment]
Althoug this feature would be very welcome, a fairly easy solution is to impose a dynamic but fixed-size limit in the query
(e.g. based on a postgres data source)
`WHERE
  $__timeFilter(ts) AND
  ts >= NOW() - interval '180' day`

**Comment by Subetov** [comment]
+1 this feature is very necessary

**Comment by pfischer64** [comment]
We'll need this feature to limit the load on the database server.

**Comment by hjc668** [comment]
same trouble, any updates?

**Comment by lfdominguez** [comment]
> Althoug this feature would be very welcome, a fairly easy solution is to impose a dynamic but fixed-size limit in the query (e.g. based on a postgres data source) `WHERE $__timeFilter(ts) AND ts >= NOW() - interval '180' day`

And if the user changes the query before sending to Grafana API because the full query is sent, I was hoping that Grafana implements something that masks the query to view-only users.

**Comment by Infant-Rashmi** [comment]
Is there any update, it a most wanted feature.

**Comment by wkneewalden** [comment]
+1

**Comment by karolina1819** [comment]
This is exactly what my team needs for Elastic Search database.

**Comment by akrea** [comment]
+1

**Comment by xhci** [comment]
+1

**Comment by serkanbudak** [comment]
+61

**Comment by pulsar666** [comment]
+1

**Comment by lingyufei** [comment]
+1

**Comment by cfradewavecom** [comment]
+1

**Comment by frjaraur** [comment]
+2

**Comment by www44** [comment]
+1

**Comment by jlynch93** [comment]
+1

**Comment by mampersat** [comment]
+1

**Comment by deryann** [comment]
+1

**Comment by life5ign** [comment]
+1000; I have semi- technical people using the panels in higher levels of QA; I need to make sure they don't start and leave many, heavy, long running queries; similar issue to "lack of this feature prevents public dashboards from being DoS'd"

**Comment by vpmartin** [comment]
+1, still needed in 2024 for publicly available dashboards.

**Comment by xhci** [comment]
> +2; I have semi- technical people using the panels in higher levels of QA; I need to make sure they don't start and leave many, heavy, long running queries; similar issue to "lack of this feature prevents public dashboards from being DoS'd"

This is exactly our use-case. Some ranged queries can be extremely expensive if not constrained.

**Comment by snuggie12** [comment]
+1 if your datasource costs a lot of money and naive users do large time ranges.

**Comment by joaolce** [comment]
+1

**Comment by matthewlisp** [comment]
+1

**Comment by StuartUSA** [comment]
+1

**Comment by chrischjh** [comment]
+1  Really needed in public dashboards

**Comment by immuslim** [comment]
+1

**Comment by srijanshukla18** [comment]
+1

**Comment by 404notfoundhard** [comment]
+1

**Comment by GustavJaner** [comment]
+1

**Comment by jcdauchy-moodys** [comment]
+1

**Comment by mvandenboogaard-tilaa** [comment]
+1

**Comment by WoCha-FR** [comment]
+1

**Comment by benoitschipper** [comment]
This would definitely be helpful in our public Dashboards

**Comment by peter-edb** [comment]
Agree, +1 was looking for this today

**Comment by theotheredbaron** [comment]
Think I mentioned this before, or it was on another similar ticket I raised.

The way I fixed it was to use MySQL VIEWs.  Since the data we were examining in grafana was all from a mysql DB I created various VIEWs of all the tables with date limits (e.g. user_comments_last_month etc) and pointed grafana at those instead.

This was quite a while ago and I can't remember the details as we moved to other software so that "trick" may not work nowadays, or on other RDBMSs.

Good luck!  The issue's been open over eight years so I doubt the grafana team has time to look at it :)

**Comment by cfradewavecom** [comment]
@theotheredbaron that does not work with other datasources like Prometheus, the way i see it this must be limited at the frontend level. But i agree with you that grafana team is not interested in this feature ($).

**Comment by theotheredbaron** [comment]
@cfradewavecom yep, I was only talking about mysql and it worked OK for that.  Management wanted a dashboard and immediately selected "since the beginning of time" or whatever it's called so I had to find something fairly quickly before the monitoring servers exploded...

**Comment by TheMatrix97** [comment]
In 2020 a PR addressing this issue was closed without further explanation..... I sincerely consider this should be implemented in Grafana, mainly to allow the enforcement of a max-range variable for read-only dashboard users.

**Comment by bastifpv** [comment]
+1

**Comment by antonblazquezd** [comment]
+1

**Comment by tmanninger** [comment]
+1

**Comment by erendogan51** [comment]
+1
